examples/optimizers/Main.hs

```diff
@@ -0,0 +1,226 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main where
+
+{- Optimizer implementations run on standard test functions -}
+
+import Control.Monad (foldM, when)
+import GHC.Generics
+import Prelude hiding (sqrt)
+import Text.Printf (printf)
+
+import Torch.Tensor
+import Torch.TensorFactories (eye', zeros')
+import Torch.Functions
+import Torch.Autograd
+import Torch.NN
+
+import TestFunctions
+
+-- Optimizers
+
+type LearningRate = Tensor
+type Gradient = [Tensor]
+
+-- gradient descent step
+gd :: LearningRate -> [Parameter] -> Gradient -> [Tensor]
+gd lr parameters gradients = zipWith step depParameters gradients
+  where
+    step p dp = p - (lr * dp)
+    depParameters = fmap toDependent parameters
+
+-- gradient descent with momentum step
+--     lr              beta      memory      parameters     gradients
+gdm :: LearningRate -> Float -> [Tensor] -> [Parameter] -> Gradient -> [(Tensor, Tensor)]
```

Comment:
this seems to be a better way to document the arguments:
```suggestion
gdm
  :: LearningRate -- ^ learning rate
  -> Float -- ^ beta
  -> [Tensor] -- ^ memory
  -> [Parameter] -- ^ parameters
  -> Gradient -- ^ gradients
  -> [(Tensor, Tensor)]
```
cf. https://haskell-haddock.readthedocs.io/en/latest/markup.html#constructors-and-record-fields

Reply:
Using this convention now in `Optimizers.hs`